streaming/src/main/scala/org/apache/spark/streaming/receiver/ReceiverSupervisor.scala

```diff
@@ -136,12 +141,19 @@ private[streaming] abstract class ReceiverSupervisor(
   def stopReceiver(message: String, error: Option[Throwable]): Unit = synchronized {
     try {
       logInfo("Stopping receiver with message: " + message + ": " + error.getOrElse(""))
-      receiverState = Stopped
-      receiver.onStop()
-      logInfo("Called receiver onStop")
-      onReceiverStop(message, error)
+      receiverState match {
+        case Initialized =>
+          logWarning("Skip stopping receiver because it has not yet stared")
+        case Started =>
```

Comment:
Thats the right way to do this. But then by that logic we should start only when the state is Init or Stopped. We are not doing that. In fact, that leads to the question of locking the state correctly and all. I dont want to get into fixing those in this PR. 

I think its better not change stopReceiver() and rather change stop() to call stopReceiver() only when the state is started